docs/Defining_Platforms.md

```diff
@@ -0,0 +1,282 @@
+
+For example, in the `reference` platform:
+* `params.h` includes `params_float.h` and `params_quint.h`
+* `Buffer.hpp` includes `FloatBuffer.hpp` and `QUInt8Buffer.hpp`
+* `intrinsics.h` includes `intrinsics_float.h` and `intrinsics_quint8.h`
+
+### Platform-Specific Constants
+
+This file must define a set of macros (compile-time constants) that are selected to tune the loop nests to various hardware capabilities (TODO: FILL IN BETTER DEFINITION).  Note that each macro must have a prefix corresponding to the data type it is used with.  For example, `FLOAT_` prefix is used for `float` types and `QUINT8_` is used for quantized unsigned 8-bit integer types.
+
+* <TYPE>_W_ob -   (FILL IN: some sort of output blocking factor)
+* <TYPE>_C_ob -   Output channel blocking factor
+* <TYPE>_C_ib -   Input channel blocking factor (currently this must be the same as the output channel blocking factor).
+* <TYPE>_SIMD -   ???
+* <TYPE>_UNROLL - ???
+* <TYPE>_NUM_FMA - ??? (is this used?)
+* <TYPE>_NUM_MAX - ??? (is this used?)
+* <TYPE>_NUM_LOAD - ??? (is this used?)
+* <TYPE>_NUM_STORE - ??? (is this used?)
+
+### Support for Buffers of different data types
+
+#### Dynamic allocation and deallocation
+
+This section is TBD:
+
+Two free functions need to be defined in a top-level Buffer.hpp file:
+
+* A factory function templated on the Buffer type:
+```
+template <class BufferT>
+inline BufferT *alloc_buffer(size_t num_elements);
+```
+* A way to deallocate the buffers created by the factory:
+```
+template <class BufferT>
+inline void free_buffer(BufferT *buffer)
+```
+
+TBD: We believe these are needed abstract away how memory is allocated for systems like the Arduino where only static allocation is allowed.
+
+#### Buffer classes for specific types
+
+Untemplated classes are used to declare a different Buffer class for each data type supported by the platform.
+
+The convention used is that FloatBuffer class is declared for floating point types, QUInt8Buffer is used for quantized, unsigned 8-bit integers.
+
+##### Platform constants (again)
+
+Each buffer class must contain static const members for all of the values defined in the corresponding `params_<type>.h` file:
+
+```
+class FloatBuffer
+{
+public:
+    static uint32_t const   W_ob{FLOAT_W_ob};
+    static uint32_t const   C_ob{FLOAT_C_ob};
+    static uint32_t const   SIMD{FLOAT_SIMD};
+    static uint32_t const UNROLL{FLOAT_UNROLL};
+    static uint32_t const   C_ib{FLOAT_C_ib};
+
+    static uint32_t const   NUM_FMA{FLOAT_NUM_FMA};
+    static uint32_t const   NUM_MAX{FLOAT_NUM_MAX};
+    static uint32_t const  NUM_LOAD{FLOAT_NUM_LOAD};
+    static uint32_t const NUM_STORE{FLOAT_NUM_STORE};
+};
+```
+
+Note that the prefix must be removed so that each `<Type>Buffer` class defines exactly the same set of constants.
+
+TBD: make sure these are compile-time constants.
```

Comment:
doesn't `const` already ensure that?

Reply:
Need to confirm either experimentally with Compiler Explorer or relevant language in the ISO C++ Specification.